### DISCUSSION TITLE
Jump-to-definition, opam libraries, and "No config found for file _opam/lib/…"

### DISCUSSION THREAD

**Discussion by ELLIOTTCABLE** [ocaml/ocaml-lsp#1062] [Q&A]
One consistent pain when working with the OCaml Platform is using jump-to-definition to browse code that's in opam dependencies:

<img width="1129" alt="image" src="https://user-images.githubusercontent.com/200/232620648-9e02b7f7-bae4-410e-9e48-be6d4c56fabc.png">

Is there any workaround to "link up" the source files that jump-to-definition hits, and the actual build-products from Dune? Some way to make the browsing/spelunking experience better?

**Comment by sidkshatriya** [comment]
Jumping to definitions in installed libraries in OCaml is a long standing pain point. This is a problem that does not exist in Rust for instance. You can always jump to definition in crates which are dependencies. This gives you full visibility into the code your project will run.

There are some discussions about this on discuss.ocaml.org that can help you for OCaml. Here is one of them:

"Merlin: Goto definition inside installed libraries "
https://discuss.ocaml.org/t/merlin-goto-definition-inside-installed-libraries/2788

Even though that discussion is about Merlin, it is applicable to your question because ocaml-lsp uses merlin as a backend for its OCaml "intelligence".

**TL;DR** the best approach that I know of currently (which works in most of the cases) is to "vendor" your dependencies. (Look up dune vendoring to learn more). So, you would NOT install the dependencies in your opam switch, rather the source code of your dependencies would be embedded with your project source code (usually under the `vendor/` subfolder or `duniverse` subfolder -- keep reading).

There is a tool that does this for you. Its called `opam monorepo`. (See https://github.com/tarides/opam-monorepo) . It vendors all your opam dependencies. However it only does vendoring for projects that work with dune which can be limiting. So the OCaml community has something called duniverse (See https://github.com/dune-universe/opam-overlays ) which simply provides dune versions of packages that don't use dune. You need to add this duniverse as an overlay repository on your OCaml switch.

Sorry if this sounds too intricate. Its actually not. Here is an example:
https://discuss.ocaml.org/t/merlin-goto-definition-inside-installed-libraries/2788/21

(I used dream as an example in the link above, the project does some complicated things with dependencies already so can be a bit tricky to use opam monorepo with. Try it with some other project. This works properly in even recent versions of OCaml 4.14.1, 5.0 etc. so don't worry about the comment being outdated).  

For completeness I would like to add that `opam-monorepo` is the most general approach that will accomplish what your want. Though sometimes you don't want to vendor _everything_ and are only interested in a couple of libraries where you want to go to the source code for. In that case, just manually dune vendor them. Dune vendoring is a really useful -- Everyone in OCaml/dune ecosystem should try it out (if they haven't already).
